### DISCUSSION TITLE
Get All terminal content

### DISCUSSION THREAD

**Discussion by bajrangCoder** [sourcelair/xterm.js#4467] [Q&A] [answered]
Hi,
I want to get all content of xtermjs terminal. To save it and whenever user will open xtermjs terminal, i want to continue from there where he left. Is it possible? If yes please answer..

**Comment by jerch** [comment]
You might want to have a look at the [serialize addon](https://github.com/xtermjs/xterm.js/tree/master/addons/xterm-addon-serialize), it preserves most of the terminal buffer state and operation modes. It returns a string, that can be fed back to your newly created terminal:
```javascript
const serializeAddon = new SerializeAddon();
oldTerminal.loadAddon(serializeAddon);
...
const terminalState = serializeAddon.serialize();
...
const newTerminal = new Terminal(...);
newTerminal.write(terminalState);
```

Sidenote: To avoid deserialization artefacts it might be needed to give the new terminal the same size as the old one at the time of serialization (simply resize the new terminal into proper shape afterwards).

**Comment by bajrangCoder** [comment]
Ok, Thanks for help
Let me try..

**Comment by bajrangCoder** [comment]
Yes , it generates all the terminal content but I am using this inside my websocket.onmessage function and trying to save all available content of terminal into a file on new message, but it doesn't save every time, but when i console log the the serializer data ( means terminal data ) then it logs data one by one.
What's the problem there and is there any solution 

I am using something like this :
```javascript
this.ws.onmessage = (event) => {
   this.$terminal.write(event.data, () => {
     let terminalState = this.$serializerAddon.serialize();
     // Code for save the terminal data in a file 
     await fs(file path).writeFile(terminalState);
   })
}
```

Edit:
after some research i found that serializer send data row wise .
Then i need to save it in a string or a array???

**Reply by jerch** [reply, marked as answer]
Unless you want to write a terminal recorder, invoking the serializer on every incoming chunk of data is def. the wrong approach. And even for a terminal recorder that is most likely not the best way to do things, as it will re-serialize higher up lines over and over creating very bad runtime.

> after some research i found that serializer send data row wise .

Not sure what you mean by that. The serializer walks the buffer cells line-wise yes, but it creates a full dump of the buffer content in a single string. The string is meant to restore most of the terminal state at the time of serialization, thus will contain multiple lines, if there were more than a single line.

Your code above looks very weird, and should not even run due to mixing synchronous code with an asynchronous await call. Plz read up promises and their impact on control flow order.

**Reply by bajrangCoder** [reply]
Ok

**Reply by bajrangCoder** [reply]
Sorry, due my a silly mistake, there is confusion arisen, Now every thing working well.

**Comment by bajrangCoder** [comment]
I have one more question.

As i have created a Linux based terminal and it's working smoothly, but with some keyboard specially with keyboard that supports Autocompletion or with some mobile keyboard, It creates problem. As when anyone enter any command with physical keyboard,etc  and hit <kbd>enter key</kbd> then that Command send to pty and output will be return and done but with some special keyboards while entering any command if i click <kbd>enter </kbd>then it will not run instead it creates a new line and then i again hit <kbd>enter</kbd> then it sends to pty and output will be returned , But if i enter my command with special keyboards and before pressing <kbd>enter key</kbd> , if i press <kbd>spacebar</kbd> and then press <kbd>enter key</kbd> then it runs as normal.
Is there any solution for this to make it normal as other keyboard?

Here are some pics of that special case:

**Before pressing enter key**
![Screenshot_20230410-224435_1](https://user-images.githubusercontent.com/71929976/230956164-b3b1ad0d-2405-44cd-9bb7-7b2c72dc219e.png)

**After pressing enter key only once**
![Screenshot_20230410-224504_1](https://user-images.githubusercontent.com/71929976/230956154-fcca0dc1-51cc-4719-8726-2fb88f2ca0da.png)


**Note: Here special keyboards represents the keyboard which supports Autocompletion or mobile keyboards like gboard , etc**

**Reply by jerch** [reply]
There are known issues with some keyboard apps under android (plz search the issue list, I dont remember the issue numbers). Background on this - those keyboard apps have a very own understanding of placing typed stuff into input fields, some of them even delete & replace previously typed characters by their auto-completion logic. Thats mostly incompatible with xterm.js atm, since xterm.js does a char by char insert on the PTY, where deleting works quite different if at all (depends on the TTY mode).

A proper fix with those keyboard apps does not exist atm, and I doubt to some degree that it is possible in a reliable way due their very creative way of juggling input characters.

As a quickfix - disable those apps for terminal input. This has another advantage - their language models for auto-completion will most likely work against your terminal input and annoy ppl. By disabling you get rid of both issues.

(Edit: There is another issue with those fancy mobile keyboards - they typically miss important keys for terminal interaction. Since I am not an android developer I dont know the the solution to this issue, maybe its time for a custom terminal keyboard app?)

**Reply by bajrangCoder** [reply]
Ok but Replit has get rid of this issue on their mobile version.

**Reply by jerch** [reply]
Then ask them, how they did that? We would gladly review such a solution, and even may integrate it, if it works reliable. But this really depends on how much of a workaround this is, as we cannot provide shims for every ancient keyboard app out there. Android is just a mess at that point as far as I get it.

**Reply by bajrangCoder** [reply]
hmm
